sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_CustomEventProcessor.md

```diff
@@ -0,0 +1,439 @@
+            }
+
+            if (lastEvent != null)
+            {
+                await CheckpointAsync(
+                    partition,
+                    lastEvent,
+                    cancellationToken)
+                .ConfigureAwait(false);
+            }
+        }
+        catch (Exception ex)
+        {
+            // It is very important that you always guard against exceptions in
+            // your handler code; the processor does not have enough
+            // understanding of your code to determine the correct action to take.
+            // Any exceptions from your handlers go uncaught by the processor and
+            // will NOT be redirected to the error handler.
+
+            Console.WriteLine($"Exception while processing events: { ex }");
+        }
+    }
+
+    protected override Task OnProcessingErrorAsync(
+        Exception exception,
+        EventProcessorPartition partition,
+        string operationDescription,
+        CancellationToken cancellationToken)
+    {
+        try
+        {
+            if (partition != null)
+            {
+                Console.Error.WriteLine(
+                    $"Exception on partition { partition.PartitionId } while " +
+                    $"performing { operationDescription }: {exception}");
+            }
+            else
+            {
+                Console.Error.WriteLine(
+                    $"Exception while performing { operationDescription }: { exception }");
+            }
+        }
+        catch (Exception ex)
+        {
+            // It is very important that you always guard against exceptions
+            // in your handler code; the processor does not have enough
+            // understanding of your code to determine the correct action to
+            // take.  Any exceptions from your handlers go uncaught by the
+            // processor and will NOT be handled in any way.
+
+            Console.WriteLine($"Exception while processing events: { ex }");
+        }
+
+        return Task.CompletedTask;
+    }
+}
+```
+
+This should look familiar if you've used `EventProcessorClient` before, but there are a few small differences you need to be aware of:
+
+- The processor works in batches, so when you construct it you need to specify the maximum number of events you'll want to process at once, using the `eventBatchMaximumCount` parameter.  The `IEnumerable<EventData>` that is passed `OnProcessingEventBatchAsync` to will never contain more items than this maximum count, but may contain fewer.
+
+- If there are fewer unprocessed messages than `eventBatchMaximumCount` the processor also allows for a maximum amount of time that it will wait for more events to become available.  By default, this is 60 seconds, but can be controlled by setting the `MaximumWaitTime` property of the `EventProcessorOptions` parameter.  If this is set to `null` it means that the processor will forever until it has received some events.
```

Comment:
@ellismg:  The corresponding statement in the blog article has an error.   Due to the way that the AMQP library works, setting a `MaximumWaitTime` will allow the processor to wait until any event has been received; it won't try to build up the `eventBatchMaximumCount` first.   Whatever was read from the service in a single operation is dispatched.

![image](https://user-images.githubusercontent.com/913445/113765470-4e978a80-96ea-11eb-91b7-541b0c9680b6.png)

Reply:
Thanks for pointing this out.  I will try to get the blog post updated.  Sorry for the incorrect assertion and thanks for helping me get my mental model right here.

Reply:
Not at all.  It's odd and unexpected behavior.   We had it documented this way for a while.  I opened a bug on the AMQP library the first time that a customer reported that they were seeing inconsistent batch sizes.